nodejs/awsx/cloudtrail/trail.ts

```diff
@@ -0,0 +1,115 @@
+
+import * as aws from "@pulumi/aws";
+import * as pulumi from "@pulumi/pulumi";
+import * as utils from "../utils";
+
+export class Trail extends pulumi.ComponentResource {
+    /**
+     * The managed CloudWatch Log Group.
+     */
+    public readonly logGroup: aws.cloudwatch.LogGroup | undefined;
+
+    /**
+     * The managed S3 Bucket.
+     */
+    public readonly bucket: aws.s3.Bucket | undefined;
+
+    /**
+     * The CloudWatch Trail.
+     */
+    public readonly trail: aws.cloudtrail.Trail;
+
+    /**
+     * Create a Trail resource with the given unique name, arguments, and options.
+     *
+     * @param name The _unique_ name of the resource.
+     * @param args The arguments to use to populate this resource's properties.
+     * @param opts A bag of options that control this resource's behavior.
+     */
+    constructor(name: string, args: TrailArgs, opts?: pulumi.CustomResourceOptions) {
+        super("aws:cloudtrail:x:Trail", name, {}, opts);
+
+        let bucketName: pulumi.Output<string>;
+        if (args.s3BucketName !== undefined) {
+            bucketName = pulumi.output(args.s3BucketName);
+        } else {
+            this.bucket = bucketWithCloudTrailPolicy(name, this);
+            bucketName = this.bucket.id;
+        }
+
+        let cloudWatchLogGroupArn: pulumi.Output<string> | undefined = undefined;
+        if (args.sendToCloudWatchLogs) {
+            this.logGroup = new aws.cloudwatch.LogGroup(name, args.cloudWatchLogGroupArgs, {parent: this});
+            cloudWatchLogGroupArn = pulumi.interpolate`${this.logGroup.arn}:*`;
+        }
+
+        const trailArgs = {
+            s3BucketName: bucketName,
+            cloudWatchLogGroupArn: cloudWatchLogGroupArn,
+            ...args,
+        };
+
+        this.trail = new aws.cloudtrail.Trail(name, trailArgs, {parent: this});
+    }
+}
+
+
+type TrailArgs = utils.Overwrite<aws.cloudtrail.TrailArgs, {
+    /**
+     * If CloudTrail pushes logs to CloudWatch Logs in addition to S3.
+     *
+     * Disabled for cost out of the box.
+     *
+     * @default false
+     */
+    readonly sendToCloudWatchLogs?: pulumi.Input<boolean>;
+
+    /**
+     * If sendToCloudWatchLogs is enabled, provide these arguments to configure
+     * the log group.
+     */
+    readonly cloudWatchLogGroupArgs?: aws.cloudwatch.LogGroupArgs;
+
+    /**
+     * Specifies the name of the S3 bucket designated for publishing log files.
+     */
+    readonly s3BucketName?: pulumi.Input<string>;
+}>;
+
+function bucketWithCloudTrailPolicy(name: string, parent: pulumi.Resource | undefined): aws.s3.Bucket {
+    const bucket = new aws.s3.Bucket(name, {}, {parent: parent});
+    const policy = new aws.s3.BucketPolicy(name, {
+        bucket: bucket.id,
+        policy: {
+            Version: "2012-10-17",
+            Statement: [{
+                Sid: "AWSCloudTrailAclCheck",
+                Effect: "Allow",
+                Principal: {Service: "cloudtrail.amazonaws.com"},
```

Comment:
What do you think about using `aws.iam.principals.CloudtrailPrincipal` here instead?